pkg/v1/tkg/test/tkgctl/shared/e2e_tkrresolver.go

```diff
@@ -0,0 +1,358 @@
+			LogOptions: tkgctl.LoggingOptions{
+				File:      filepath.Join(logsDir, clusterName+".log"),
+				Verbosity: input.E2EConfig.TkgCliLogLevel,
+			},
+		})
+
+		Expect(err).To(BeNil())
+	})
+
+	It("Should create workload cluster with default kubernetes version and verify infra machine images are resolved correctly", func() {
+		By(fmt.Sprintf("Creating a workload cluster %q", clusterName))
+
+		options := framework.CreateClusterOptions{
+			ClusterName:  clusterName,
+			Namespace:    namespace,
+			Plan:         "dev",
+			CniType:      input.Cni,
+			OtherConfigs: input.OtherConfigs,
+		}
+		if input.Plan != "" {
+			options.Plan = input.Plan
+		}
+
+		if input.E2EConfig.InfrastructureName == "vsphere" {
+			if input.Cni == "antrea" {
+				if clusterIP, ok := os.LookupEnv("CLUSTER_ENDPOINT_ANTREA"); ok {
+					options.VsphereControlPlaneEndpoint = clusterIP
+				}
+			}
+			if input.Cni == "calico" {
+				if clusterIP, ok := os.LookupEnv("CLUSTER_ENDPOINT_CALICO"); ok {
+					options.VsphereControlPlaneEndpoint = clusterIP
+				}
+			}
+		}
+
+		clusterConfigFile, err := framework.GetTempClusterConfigFile(input.E2EConfig.TkgClusterConfigPath, &options)
+		Expect(err).To(BeNil())
+
+		defer os.Remove(clusterConfigFile)
+
+		err = tkgCtlClient.CreateCluster(tkgctl.CreateClusterOptions{
+			ClusterConfigFile: clusterConfigFile,
+			Edition:           "tkg",
+			Namespace:         namespace,
+		})
+		Expect(err).To(BeNil())
+
+		By(fmt.Sprintf("Validating the TKR data after cluster %q is created", clusterName))
+		verifyTKRData(context, mcProxy, options.ClusterName, options.Namespace)
+
+		By(fmt.Sprintf("Deleting workload cluster %q", clusterName))
+		err = tkgCtlClient.DeleteCluster(tkgctl.DeleteClustersOptions{
+			ClusterName: clusterName,
+			Namespace:   namespace,
+			SkipPrompt:  true,
+		})
+		Expect(err).To(BeNil())
+
+		By("Test successful !")
+	})
+
+	It("Should upgrade the cluster and verify infra machine images are resolved correctly", func() {
+		By(fmt.Sprintf("Creating a workload cluster %q", clusterName))
+
+		tkgBOMConfigClient := tkgconfigbom.New(input.E2EConfig.TkgConfigDir, nil)
+		defaultTKRVersion, err := tkgBOMConfigClient.GetDefaultTKRVersion()
+
+		Expect(err).ToNot(HaveOccurred(), "failed to get the default TKR version")
+
+		var defaultTKR, oldTKR *runv1.TanzuKubernetesRelease
+		Eventually(func() bool {
+			tkrs := mcProxy.GetTKRs(context)
+			defaultTKR, oldTKR = getTKRsForUpgrade(defaultTKRVersion, tkrs)
+			return defaultTKR != nil && oldTKR != nil
+		}, waitTimeout, pollingInterval).Should(BeTrue(), "failed to get at least 2 TKRs(upgradable) to perform upgrade tests")
+
+		input.E2EConfig.KubernetesVersionOld = oldTKR.Spec.Kubernetes.Version
+
+		options := framework.CreateClusterOptions{
+			ClusterName:       clusterName,
+			Namespace:         namespace,
+			Plan:              "dev",
+			CniType:           input.Cni,
+			KubernetesVersion: input.E2EConfig.KubernetesVersionOld,
```

Comment:
If we do the above (factor out the common prep code into `BeforeEach()`), we could simply assign this as:
```go
options.KubernetesVersion = input.E2EConfig.KubernetesVersionOld
```